docs/datastorage/datafixers.md

```diff
@@ -0,0 +1,107 @@
+For the example above, the walker would look through `compound` for each ItemStack, and then call `fixer.process(FixTypes.ITEM_INSTANCE, stackNBT)` on it, which will recursively invoke all ITEM_INSTANCE fixers and walkers.
+
+Registering Your Own Fixes
+--------------------------
+
+### Registering a fix
+1. Obtain the `DataFixer` using `FMLCommonHandler.instance().getDataFixer()`.
+2. Call `fixer.init("modid", modDataVersion)`, and save the `ModFixs` return value.
+3. Call `ModFixs.registerFix(IFixType type, IFixableData fixer)` to register your fix.
+
+### Registering a walker
+1. Obtain the `DataFixer` using `FMLCommonHandler.instance().getDataFixer()`
+2. Call `fixer.registerWalker(FixTypes type, IDataWalker walker)` to register your walker.
+
+Fixer Case Study: Flattening
+----------------------------
+
+An example use case of a fixer is to implement the [flattening][], where different items formerly expressed as metadata variants of a single ID are being split out into their own ID's, e.g. `minecraft:dye` meta 0 -> `minecraft:black_dye`. This case is not easily covered by FML's `MissingMappingsEvent`, since that only supports wholesale remapping of *all* metadata variants of the old item to the same meta of the new item.
+This issue would be resolved in the following manner:
+
+```Java
+IFixableData flattener = new IFixableData() {
+	private String getNewID(String oldID, int meta) { /* omitted */ }
+
+	@Override
+	public int getFixVersion() {
+		return 5; // run this fixer on all discovered ItemStacks with lower version than 5
+	}
+
+	@Override
+	public NBTTagCompound fixTagCompound(NBTTagCompound compound) {
+		// Convert to new ID
+		compound.setString(getNewID(compound.getString("id"), compound.getInteger("Damage")));
+	}
+};
+
+// Register to ITEM_INSTANCE so the walker is called for all encountered ItemStacks
+ModFixs modfixes = FMLCommonHandler.instance().getDataFixer().init("mymod", 5);
+modfixes.registerFix(FixTypes.ITEM_INSTANCE, flattener);
+```
+
+Walker Case Study: Dolly
+------------------------
+
+Mods must take care to call the Fixer system whenever they hold onto NBT data of vanilla or other mods.
+For example, consider a dolly item that allows one to store any Tile Entity into itself, under the tag "StoredTE".
+If the owner of that Tile Entity registers a `FixTypes.BLOCK_ENTITY` fixer to run on that Tile Entity, the DataFixer system is not aware that your dolly item has a Tile Entity tag stored inside of itself and thus the dolly is left with an invalid tag inside of it.
+This issue would be resolved in the following manner:
+
+```Java
+IDataWalker dollyWalker = new IDataWalker() {
+	@Override
+	public NBTTagCompound process(IDataFixer fixer, NBTTagCompound compound, int versionIn) {
+		// hasKey checks omitted here, get the stored TE
```

Comment:
Should we assume the process is called systematically on all items and it's up to mod author to verify that this ItemStack is its own?
Where's the version coming from if it's applied systematically?

Reply:
yes, you have to check the id tags in the stack to make sure it's your item

not sure what your second question means, but refer to my earlier comment

Reply:
So the version is the one attached to your mod when the original data was saved, as found in the chunk NBT? in which case, register method will actually attach the walker instance to the calling mod instance, so later on the version parameter is the one matching that specific mod.
Which means one mod cannot fix/patch another mod that way?

Reply:
I'm not sure what you're asking but yes in general mods aren't aware of the version of other mods or vanilla.